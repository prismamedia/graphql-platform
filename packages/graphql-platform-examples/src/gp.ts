import { GraphQLPlatform } from '@prismamedia/graphql-platform';
import { config as Article } from './resources/article';

export const gp = new GraphQLPlatform({
  resources: {
    Article,
  },
});

console.debug(gp.schema);
