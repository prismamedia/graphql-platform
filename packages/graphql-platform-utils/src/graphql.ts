import * as graphql from 'graphql';
import { getOptionalFlag, OptionalFlag } from './config.js';
import { getEnumKeys } from './enum.js';
import { UnexpectedValueError } from './error.js';
import { indefinite } from './indefinite.js';
import { isNil, type Nillable } from './nil.js';
import type { Path } from './path.js';
import { isPlainObject, type PlainObject } from './plain-object.js';

export const isGraphQLResolveInfo = (
  maybeGraphQLResolveInfo: unknown,
): maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo =>
  isPlainObject(maybeGraphQLResolveInfo) &&
  typeof maybeGraphQLResolveInfo['fieldName'] === 'string' &&
  Array.isArray(maybeGraphQLResolveInfo['fieldNodes']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['returnType']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['parentType']);

export function assertGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
  path?: Path,
): asserts maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  if (!isGraphQLResolveInfo(maybeGraphQLResolveInfo)) {
    throw new UnexpectedValueError(
      `a GraphQLResolveInfo`,
      maybeGraphQLResolveInfo,
      { path },
    );
  }
}

export const isGraphQLASTNode = <TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
): maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] =>
  isPlainObject(maybeGraphQLASTNode) && maybeGraphQLASTNode['kind'] === kind;

export function assertGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
  path?: Path,
): asserts maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  if (!isGraphQLASTNode(maybeGraphQLASTNode, kind)) {
    throw new UnexpectedValueError(`a GraphQL ${kind}`, maybeGraphQLASTNode, {
      path,
    });
  }
}

export const parseGraphQLScalarValue = <TInternal>(
  type: graphql.GraphQLScalarType<TInternal, any>,
  maybeScalarValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeScalarValue)) {
    return maybeScalarValue;
  }

  try {
    return type.parseValue(maybeScalarValue);
  } catch (error) {
    throw new UnexpectedValueError(indefinite(type.name), maybeScalarValue, {
      cause: error,
      path,
    });
  }
};

export const parseGraphQLEnumValue = <TInternal = any>(
  type: graphql.GraphQLEnumType,
  maybeEnumValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeEnumValue)) {
    return maybeEnumValue;
  }

  const enumValue = type
    .getValues()
    .find(({ value }) => value === maybeEnumValue);

  if (!enumValue) {
    throw new UnexpectedValueError(
      `${indefinite(type.name)} (= a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}")`,
      maybeEnumValue,
      { path },
    );
  }

  return enumValue.value;
};

export const parseGraphQLLeafValue = (
  type: graphql.GraphQLLeafType,
  maybeLeafValue: unknown,
  path?: Path,
): any => {
  if (isNil(maybeLeafValue)) {
    return maybeLeafValue;
  }

  return graphql.isScalarType(type)
    ? parseGraphQLScalarValue(type, maybeLeafValue, path)
    : parseGraphQLEnumValue(type, maybeLeafValue, path);
};

export type CreateGraphQLEnumTypeOptions = {
  description?: Nillable<string>;
  useKeyAsValue?: OptionalFlag;
};

export const createGraphQLEnumType = (
  name: string,
  enumerable: PlainObject,
  options?: CreateGraphQLEnumTypeOptions,
): graphql.GraphQLEnumType =>
  new graphql.GraphQLEnumType({
    name,
    description: options?.description,
    values: Object.fromEntries(
      getEnumKeys(enumerable).map<[string, graphql.GraphQLEnumValueConfig]>(
        (key) => [
          key,
          {
            value: getOptionalFlag(options?.useKeyAsValue, false)
              ? undefined
              : enumerable[key],
          },
        ],
      ),
    ),
  });
