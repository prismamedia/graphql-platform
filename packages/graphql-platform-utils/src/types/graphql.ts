export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
