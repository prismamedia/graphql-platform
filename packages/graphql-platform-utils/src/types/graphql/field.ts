import {
  GraphQLFieldConfig as GraphQLFieldConfigBase,
  GraphQLResolveInfo,
} from 'graphql';
import { MaybePromise } from '../maybe-promise';

export type FieldResolverArgs<TSource = any, TContext = any, TArgs = any> = [
  TSource,
  TArgs,
  TContext,
  GraphQLResolveInfo,
];

/** Override the original "GraphQLFieldResolver" to be able to define the result */
export type FieldResolver<
  TSource = any,
  TContext = any,
  TArgs = any,
  TResult = any
> = (
  ...args: FieldResolverArgs<TSource, TContext, TArgs>
) => MaybePromise<TResult>;

/** Override the original "GraphQLFieldConfig" to be able to define the resolver's result */
export interface FieldConfig<
  TSource = any,
  TContext = any,
  TArgs = any,
  TResult = any
> extends Omit<GraphQLFieldConfigBase<TSource, TContext, TArgs>, 'resolve'> {
  resolve?: FieldResolver<TSource, TContext, TArgs, TResult>;
}

export interface FieldConfigMap<TSource = any, TContext = any, TArgs = any> {
  [key: string]: FieldConfig<TSource, TContext, TArgs>;
}
