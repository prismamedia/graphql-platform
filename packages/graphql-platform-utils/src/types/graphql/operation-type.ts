// cf: https://graphql.github.io/graphql-spec/draft/#OperationType
export enum OperationType {
  Mutation = 'mutation',
  Query = 'query',
  Subscription = 'subscription',
}
