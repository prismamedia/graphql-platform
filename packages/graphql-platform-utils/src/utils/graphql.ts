export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
