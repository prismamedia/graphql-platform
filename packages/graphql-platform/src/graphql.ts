import type * as graphql from 'graphql';
import type { ConnectorInterface } from './connector-interface.js';
import type { GraphQLPlatform } from './index.js';

export type GPBoundGraphQLFieldResolver<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> = (
  this: GraphQLPlatform<TRequestContext, TConnector>,
  source: TSource,
  args: TArgs,
  context: TRequestContext,
  info: graphql.GraphQLResolveInfo,
) => TResult;

export interface GPBoundGraphQLFieldConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> extends graphql.GraphQLFieldConfig<TSource, TRequestContext, TArgs> {
  resolve?: GPBoundGraphQLFieldResolver<
    TRequestContext,
    TConnector,
    TSource,
    TArgs,
    TResult
  >;
}
