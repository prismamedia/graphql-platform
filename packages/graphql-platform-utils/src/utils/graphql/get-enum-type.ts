import { GraphQLEnumType, GraphQLEnumValueConfigMap } from 'graphql';
import { POJO } from '../../types/pojo';

export function getGraphQLEnumType(name: string, enumerable: POJO, useKeyAsValue: boolean = false): GraphQLEnumType {
  return new GraphQLEnumType({
    name,
    values: Object.entries(enumerable).reduce(
      (values: GraphQLEnumValueConfigMap, [key, value]) =>
        !Number.isNaN(Number.parseInt(key))
          ? values
          : Object.assign(values, { [key]: { value: useKeyAsValue ? key : value } }),
      Object.create(null),
    ),
  });
}
