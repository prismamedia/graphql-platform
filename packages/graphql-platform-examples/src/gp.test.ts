import { printSchema } from 'graphql';
import { gp } from './gp';

describe('GraphQL Platform', () => {
  it('generates a valid GraphQL Schema', () => {
    expect(
      printSchema(gp.schema, { commentDescriptions: true }),
    ).toMatchSnapshot();
  });
});
