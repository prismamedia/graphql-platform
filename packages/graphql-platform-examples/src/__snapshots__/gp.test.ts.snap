// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Platform generates a valid GraphQL Schema 1`] = `
"schema {
  query: query
  mutation: mutation
}

type query {
  # Gets the number of \\"Article\\" nodes
  articleCount(where: ArticleWhereInput): NonNegativeInt!

  # Retrieves a list of \\"Article\\" nodes
  articles(where: ArticleWhereInput, orderBy: [ArticleOrderByInput!], skip: NonNegativeInt, first: NonNegativeInt!): [Article!]!
}

# Integers that will have a value of 0 or more.
scalar NonNegativeInt

input ArticleWhereInput {
  id: UUID
  id_not: UUID
  id_in: [UUID!]
  id_not_in: [UUID!]
  title: NonEmptyTrimmedString
  title_not: NonEmptyTrimmedString
  title_in: [NonEmptyTrimmedString!]
  title_not_in: [NonEmptyTrimmedString!]
  slug: NonEmptyTrimmedString
  slug_not: NonEmptyTrimmedString
  slug_in: [NonEmptyTrimmedString!]
  slug_not_in: [NonEmptyTrimmedString!]
  AND: [ArticleWhereInput!]
  OR: [ArticleWhereInput!]
  NOT: ArticleWhereInput
}

# A field whose value is a generic Universally Unique Identifier: https://en.wikipedia.org/wiki/Universally_unique_identifier.
scalar UUID

# A string in which the leading and trailing whitespace characters are removed, plus it cannot be empty.
scalar NonEmptyTrimmedString

type Article {
  id: UUID!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
}

enum ArticleOrderByInput {
  id_ASC
  id_DESC
  title_ASC
  title_DESC
  slug_ASC
  slug_DESC
}

type mutation {
  # Creates a(n) \\"Article\\" node then returns it
  createArticle(data: ArticleCreateInput!): Article!
}

# Data provided to create a(n) \\"Article\\" node
input ArticleCreateInput {
  id: UUID!
  title: NonEmptyTrimmedString!
  slug: NonEmptyTrimmedString!
}
"
`;
